impl/maven-impl/src/main/java/org/apache/maven/impl/model/DefaultModelProcessor.java

```diff
@@ -94,50 +97,45 @@ public Path locateExistingPom(Path projectDirectory) {
         return pom;
     }
 
+    private Path doLocateExistingPom(Path project) {
+        if (project == null) {
+            project = Paths.get(System.getProperty("user.dir"));
+        } else if (Files.isDirectory(project)) {
+            Path pom = project.resolve("pom.xml");
+            return Files.isRegularFile(pom) ? pom : null;
+        }
+        return project;
+    }
+
     @Override
     public Model read(XmlReaderRequest request) throws IOException {
-        Objects.requireNonNull(request, "source cannot be null");
-        Path pomFile = request.getPath();
+        Path pomFile = requireNonNull(request, "source cannot be null").getPath();
         if (pomFile != null) {
-            Path projectDirectory = pomFile.getParent();
             List<ModelParserException> exceptions = new ArrayList<>();
             for (ModelParser parser : modelParsers) {
                 try {
-                    Optional<Model> model =
-                            parser.locateAndParse(projectDirectory, Map.of(ModelParser.STRICT, request.isStrict()));
-                    if (model.isPresent()) {
-                        return model.get().withPomFile(pomFile);
+                    Optional<Model> parent = parent(request, parser, pomFile.getParent());
+                    if (parent.isPresent()) {
+                        return parent.get().withPomFile(pomFile);
                     }
                 } catch (ModelParserException e) {
                     exceptions.add(e);
```

Comment:
is it mandatory `to continue if an error occurs` to give full error report? Or can we have here early exit if the branch kicks in like above in its counterpart?